src/algebra/module/pid.lean

```diff
@@ -0,0 +1,288 @@
+variables [is_dedekind_domain R]
+open unique_factorization_monoid
+
+lemma is_internal_prime_power_torsion_of_is_torsion_by_ideal {I : ideal R} (hI : I ≠ 0)
+  (hM : module.is_torsion_by_set R M I) :
+  ∃ (P : finset $ ideal R) [decidable_eq P] [∀ p ∈ P, prime p] (e : P → ℕ),
+  by exactI direct_sum.is_internal (λ p : P, torsion_by_set R M (p ^ e p : ideal R)) :=
+begin
+  classical,
+  let P := factors I,
+  have prime_of_mem := λ p (hp : p ∈ P.to_finset), prime_of_factor p (multiset.mem_to_finset.mp hp),
+  refine ⟨P.to_finset, infer_instance, prime_of_mem, λ i, P.count i, _⟩,
+  apply @torsion_is_internal _ _ _ _ _ _ _ (λ p, p ^ P.count p) _,
+  { intros p hp q hq pq, dsimp,
+    rw irreducible_pow_sup,
+    { suffices : (normalized_factors _).count p = 0,
+      { rw [this, zero_min, pow_zero, ideal.one_eq_top] },
+      { rw [multiset.count_eq_zero, normalized_factors_of_irreducible_pow
+          (prime_of_mem q hq).irreducible, multiset.mem_repeat],
+        exact λ H, pq $ H.2.trans $ normalize_eq q } },
+    { rw ← ideal.zero_eq_bot, apply pow_ne_zero, exact (prime_of_mem q hq).ne_zero },
+    { exact (prime_of_mem p hp).irreducible } },
+  { convert hM,
+    rw [← finset.inf_eq_infi, is_dedekind_domain.inf_prime_pow_eq_prod,
+      ← finset.prod_multiset_count, ← associated_iff_eq],
+    exact factors_prod hI,
+    { exact prime_of_mem }, { exact λ _ _ _ _ ij, ij } }
+end
+
+theorem is_internal_prime_power_torsion [module.finite R M] (hM : module.is_torsion R M) :
+  ∃ (P : finset $ ideal R) [decidable_eq P] [∀ p ∈ P, prime p] (e : P → ℕ),
+  by exactI direct_sum.is_internal (λ p : P, torsion_by_set R M (p ^ e p : ideal R)) :=
+begin
+  obtain ⟨I, hI, hM'⟩ := is_torsion_by_ideal_of_finite_of_is_torsion hM,
+  exact is_internal_prime_power_torsion_of_is_torsion_by_ideal hI hM'
+end
+
+end dedekind_domain
+
+variables [is_principal_ideal_ring R]
+
+theorem is_internal_prime_power_torsion_of_pid [module.finite R M] (hM : module.is_torsion R M) :
+  ∃ (ι : Type u) [fintype ι] [decidable_eq ι] (p : ι → R) [∀ i, irreducible $ p i] (e : ι → ℕ),
+  by exactI direct_sum.is_internal (λ i, torsion_by R M $ p i ^ e i) :=
+begin
+  obtain ⟨P, dec, hP, e, this⟩ := is_internal_prime_power_torsion hM,
+  refine ⟨P, infer_instance, dec, λ p, is_principal.generator (p : ideal R), _, e, _⟩,
+  { rintro ⟨p, hp⟩,
+    haveI := ideal.is_prime_of_prime (hP p hp),
+    exact (is_principal.prime_generator_of_is_prime p (hP p hp).ne_zero).irreducible },
+  { convert this, ext p : 1,
+    rw [← torsion_by_span_singleton_eq, ideal.submodule_span_eq, ← ideal.span_singleton_pow,
+      ideal.span_singleton_generator] }
+end
+
+section p_torsion
+variables {p : R} (hp : irreducible p) (hM : module.is_torsion' M (submonoid.powers p))
+variables [dec : Π x : M, decidable (x = 0)]
+
+open ideal submodule.is_principal
+include dec
+
+include hp hM
+lemma torsion_of_eq_span_pow_p_order (x : M) :
+  torsion_of R M x = span {p ^ p_order hM x} :=
+begin
+  dunfold p_order,
```

Comment:
Does the proof work without this line?

Reply:
I need this line because otherwise there's messy stuff happening with the props `nat.find` is applied to it seems